Automattic/wp-calypso
# Checkout: Hide contact validation errors during auto-completion, with external state

#### Proposed Changes

When checkout loads, it fetches saved contact details from an API endpoint, pre-fills the contact form with those details, and attempts to auto-complete the contact details step. Since https://github.com/Automattic/wp-calypso/pull/64344, this process uses the same logic as if a user filled out the form and clicked "Continue" manually. 

However, there is a downside of this approach: the contact details are validated every time someone clicks "Continue" and since this uses the same process, if the saved details are invalid the user will see an error reported as soon as checkout loads. Since the user didn't actually take any action at this point, this can be very confusing.

<img width="977" alt="validation-error" src="https://user-images.githubusercontent.com/2036909/201235660-6c218f93-0ae4-418d-bb69-29e953fe404f.png">

In this PR, we refactor the placement of the auto-completion code so that its validation call will not report any errors on failure. This is done by moving the auto-completion into the parent component of the `CheckoutStepGroup` where the validation function is run and refusing to show a validation error message if auto-completion is still in-progress.

Fixes https://github.com/Automattic/wp-calypso/issues/69211

#### Differences compared to the previous version of this PR

This is the second attempt at solving this issue after https://github.com/Automattic/wp-calypso/pull/69943 was reverted in https://github.com/Automattic/wp-calypso/pull/70721 because there was a bug with free purchases. 

In the previous PR, we moved the checkout contact details step into a wrapper component due to a leaky abstraction. There is one case where no contact details step is needed: when the purchase is free ($0 total, not full credits). Before https://github.com/Automattic/wp-calypso/pull/69943, we did not render the step in this case, but after that PR, we _did_ render the step, except that the wrapper returned null. The result was that, if the purchase was free, checkout would think that the contact details step was present and _active_ even though it was null, effectively blocking checkout.

This PR performs additional refactoring to remove the leaky abstraction and gets rid of the wrapper. In addition, https://github.com/Automattic/wp-calypso/pull/70776 adds a test to prevent the bug from happening again.

Some of the changes originally made for this PR have been merged separately with https://github.com/Automattic/wp-calypso/pull/70776 (which improves the tests) and https://github.com/Automattic/wp-calypso/pull/70778 (which refactors checkout steps to externalize the store).

#### Testing Instructions

There are unit tests for this behavior but you can test it manually as follows:

First test that successful validation still works:

- Make sure you have valid cached _domain_ contact details for a user (using a domain product will cause more fields to be visible). You can do this by visiting checkout when a domain product is in your cart, filling out the contact information step and pressing "Continue". Assuming they are valid, they will be saved for future visits to checkout.
- Add a _domain_ product to your cart and visit checkout (or reload if you are already there).
- Verify that you see the contact information step load before it is auto-completed.
- Verify that the active step is now the payment method step.

Next test that manual validation displays error messages:

- Add a _domain_ product to your cart and visit checkout.
- Click to edit the contact information step if it is not already active.
- Enter invalid contact information (eg: by clearing the "Last name" field).
- Press "Continue".
- Verify that you see a validation error.
- Verify that the active step is still the contact information step.

Finally, test that auto-completion does not display error messages:

- Apply D92139-code and sandbox the API. This will break the domain contact details validation endpoint.
- Add a _domain_ product to your cart and visit checkout.
- Verify that you see the contact information step load before it is auto-completed.
- Verify that you _do not see_ a validation error.
- Verify that the active step is still the contact information step.
